src/Chapter3.hs

```diff
@@ -480,6 +518,46 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+newtype CastleName = CastleName String
+data Castle = NoCastle | Castle CastleName | CastleAndWall CastleName
+
+data ChurchOrLibrary = Church | Library
+
+newtype House = House Inhabitants
+data Inhabitants = One | Two | Three | Four
+
+data City = City {
+  cityCastle :: Castle,
+  cityChurchOrLibrary :: ChurchOrLibrary,
+  cityHouses :: [House]
+}
+
+buildCastle :: CastleName -> City -> City
+buildCastle name city = city { cityCastle = buildCastle' name (cityCastle city)}
+  where
+    buildCastle' :: CastleName -> Castle -> Castle
+    buildCastle' n (CastleAndWall _) = CastleAndWall n
+    buildCastle' n _ = Castle n
+
+buildHouse :: House -> City -> City
+buildHouse house city = city { cityHouses = house : cityHouses city }
+
+buildWalls :: City -> City
+buildWalls city@(City NoCastle _ _) = city
+buildWalls city@(City (CastleAndWall _) _ _) = city
+buildWalls city@(City (Castle n) _ _)
+  | countInhabitants city < 10 = city
+  | otherwise = city { cityCastle = CastleAndWall n }
+
+countInhabitants :: City -> Int
+countInhabitants (City _ _ houses) = sum $ inhabitants <$> houses
+
+inhabitants :: House -> Int
+inhabitants (House One) = 1
+inhabitants (House Two) = 2
+inhabitants (House Three) = 3
+inhabitants (House Four) = 4
```

Comment:
In such situations, you can get rid of some code duplications by using `case-of` instead. Something like:

```haskell
inhabitants (House house) = case houst of
    One -> ...
```

Though, of course, you can derive `Enum` and get this function almost for free. But in this exercise we want people to practice pattern matching :slightly_smiling_face:

Reply:
Good point!